From: Stanislav Malyshev <stas@php.net>
Date: Mon Jul 18 22:20:45 2016 -0700
Subject: [PATCH] Partial fix for bug #72613 - do not treat negative returns from bz2 as size_t

5.4 backport: https://github.com/shivammathur/php-src-backports/commit/35cff0355c063323b3cea6cca6cf6c176f78486e
Binary test fixtures not included: ext/bz2/tests/72613.bz2
---
diff --git a/ext/bz2/bz2.c b/ext/bz2/bz2.c
--- a/ext/bz2/bz2.c
+++ b/ext/bz2/bz2.c
@@ -137,15 +137,20 @@ struct php_bz2_stream_data_t {
 static size_t php_bz2iop_read(php_stream *stream, char *buf, size_t count TSRMLS_DC)
 {
 	struct php_bz2_stream_data_t *self = (struct php_bz2_stream_data_t *) stream->abstract;
-	size_t ret;
+	int bz2_ret;
 	
-	ret = BZ2_bzread(self->bz_file, buf, count);
+	bz2_ret = BZ2_bzread(self->bz_file, buf, count);
 
-	if (ret == 0) {
+	if (bz2_ret < 0) {
 		stream->eof = 1;
+		return -1;
 	}
 
-	return ret;
+	if (bz2_ret == 0) {
+		stream->eof = 1;
+	}
+
+	return (size_t)bz2_ret;
 }
 
 static size_t php_bz2iop_write(php_stream *stream, const char *buf, size_t count TSRMLS_DC)
diff --git a/ext/bz2/tests/bug72613.phpt b/ext/bz2/tests/bug72613.phpt
new file mode 100644
--- /dev/null
+++ b/ext/bz2/tests/bug72613.phpt
@@ -0,0 +1,23 @@
+--TEST--
+Bug #72613 (Inadequate error handling in bzread())
+--SKIPIF--
+<?php if (!extension_loaded("bz2")) print "skip"; ?>
+--FILE--
+<?php
+$fp = bzopen(__DIR__.'/72613.bz2', 'r');
+if ($fp === FALSE) {
+    exit("ERROR: bzopen()");
+}
+$data = "";
+while (!feof($fp)) {
+    $res = bzread($fp);
+    if ($res === FALSE) {
+        exit("ERROR: bzread()");
+    }
+    $data .= $res;
+}
+bzclose($fp);
+?>
+DONE
+--EXPECT--
+DONE
\ No newline at end of file
